File: python/dynamic_programming/problem_0213_house_robber_ii.py
import unittest

class Solution:
    @staticmethod
    def rob(nums):
        # since the first and last houses are adjacent (due to being arranged in a circle), all houses will be adjacent to each other if there are 3 or less houses. we can only rob one house
        if len(nums) <= 3:
            return max(nums)
        
        # dp array starting at the first house (nums[0])
        start0 = [nums[0], nums[1]]
        for i in range(2, len(nums)-1): # excludes the very last house since it would be adjacent to the starting house (nums[0])
            start0.append(nums[i] + max(start0[:-1])) # use [:-1] to exclude the immediately adjacent house

        # dp array starting at the second house (nums[1])
        start1 = [nums[1], nums[2]]
        for i in range(3, len(nums)): # excludes first house (nums[0]) but now includes the very last house
            start1.append(nums[i] + max(start1[:-1]))

        return max(start0[-1], start0[-2], start1[-1], start1[-2])
    
"""
Basically the same as House Robber 1 but done twice to account for different starting points.
Since the last house is the neighbor of the first house, we go through the array twice, one which hits the 
very first house but excludes the last house. The other which starts on the second house but hits the very last house.
After setting up those parameters, proceed as normal like in House Robber 1. 

Time = O(n) linear dp. Iterate through the array once. Technically O(2n) 🤓
Space = O(n) for the two extra arrays. Technically O(2n)
"""
    
class TestCase(unittest.TestCase):
    def setUp(self):
        self.solution = Solution()

    def testRob1(self):
        self.testCase1 = [2,3,2]
        self.testCase1ExpectedResult = 3
        self.assertEqual(self.solution.rob(self.testCase1), self.testCase1ExpectedResult)
    def testRob2(self):
        self.testCase2 = [1,2,3,1]
        self.testCase2ExpectedResult = 4
        self.assertEqual(self.solution.rob(self.testCase2), self.testCase2ExpectedResult)
    def testRob3(self):
        self.testCase3 = [1,2,3]
        self.testCase3ExpectedResult = 3
        self.assertEqual(self.solution.rob(self.testCase3), self.testCase3ExpectedResult)

    def testRob4(self):
        self.testCase4 = [1, 1, 1]
        self.testCase4ExpectedResult = 1
        self.assertEqual(self.solution.rob(self.testCase4), self.testCase4ExpectedResult)

    def testRob5(self):
        self.testCase5 = [2, 7, 9, 3, 1]
        self.testCase5ExpectedResult = 11
        self.assertEqual(self.solution.rob(self.testCase5), self.testCase5ExpectedResult)

    def testRob6(self):
        self.testCase6 = [2, 1, 1, 2]
        self.testCase6ExpectedResult = 3
        self.assertEqual(self.solution.rob(self.testCase6), self.testCase6ExpectedResult)

    def testRob7(self):
        self.testCase7 = [1, 2, 3, 4, 5, 6, 7, 8]
        self.testCase7ExpectedResult = 20
        self.assertEqual(self.solution.rob(self.testCase7), self.testCase7ExpectedResult)

    def testRob8(self):
        self.testCase8 = [1, 1, 1, 1, 1, 1]
        self.testCase8ExpectedResult = 3
        self.assertEqual(self.solution.rob(self.testCase8), self.testCase8ExpectedResult)

if __name__ == "__main__":
    unittest.main()

File: JavaMaven/src/test/java/dynamic_programming_1d/problem_0213_house_robber_ii.java
/*
 * https://leetcode.com/problems/house-robber-ii/
 * Explanation:
 * 
 * Bottom up DP Problem. It is just House Robber 1 and we call it twice, 1 going
 * from 0 to n-2, and the other 1 to n-2
 * 
 * Time: O(n)
 * Space: O(n) or O(1)
 */

package dynamic_programming_1d;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.Test;

public class problem_0213_house_robber_ii {
    public int rob(int[] nums) {
        int n = nums.length;
        if (n == 0) return 0;
        if (n == 1) return nums[0];
        if (n == 2) return Math.max(nums[0], nums[1]);

        return Math.max(helperHouseRobberI(nums, 0, n - 2), helperHouseRobberI(nums, 1, n - 1));
    }

    public int robSpace(int[] nums) {
        int n = nums.length;
        if (n == 0) return 0;
        if (n == 1) return nums[0];
        if (n == 2) return Math.max(nums[0], nums[1]);

        return Math.max(helperHouseRobberISpace(nums, 0, n-2), helperHouseRobberISpace(nums, 1, n-1));
    }

    public int helperHouseRobberI(int[] nums, int start, int end) {
        int[] dp = new int[end - start + 1];
        dp[0] = nums[start];
        dp[1] = Math.max(nums[start], nums[start + 1]);

        for (int i = 2; i < dp.length; i++) {
            dp[i] = Math.max(nums[start + i] + dp[i - 2], dp[i - 1]);
        }
        return dp[dp.length - 1];
    }

    public int helperHouseRobberISpace(int[] nums, int start, int end) {
        int prevRob = 0;
        int prevNotRob = 0;
        int rob = 0;
        int notRob = 0;
        for (int i = start; i <= end; i++) {
            rob = prevNotRob + nums[i];
            notRob = Math.max(prevRob, prevNotRob);
            prevNotRob = notRob;
            prevRob = rob;
        }
        return Math.max(rob, notRob);
    }

    @Test
    public void testRob() {
        int[] testCase1 = { 2, 3, 2 };
        int[] testCase2 = { 1, 2, 3, 1 };
        int[] testCase3 = { 1, 2, 3, 4, 5, 1, 2, 3, 4, 5 };
        int[] testCase4 = { 5, 4, 3, 2, 1 };
        int[] testCase5 = { 1, 1, 2, 1 };
        int[] testCase6 = { 2, 7, 9, 3, 1 };
        
        int testCase1ExpectedResult = 3;
        int testCase2ExpectedResult = 4;
        int testCase3ExpectedResult = 16;
        int testCase4ExpectedResult = 8;
        int testCase5ExpectedResult = 3;
        int testCase6ExpectedResult = 11;
    
        assertEquals(testCase1ExpectedResult, rob(testCase1));
        assertEquals(testCase2ExpectedResult, rob(testCase2));
        assertEquals(testCase3ExpectedResult, rob(testCase3));
        assertEquals(testCase4ExpectedResult, rob(testCase4));
        assertEquals(testCase5ExpectedResult, rob(testCase5));
        assertEquals(testCase6ExpectedResult, rob(testCase6));

        assertEquals(testCase1ExpectedResult, robSpace(testCase1));
        assertEquals(testCase2ExpectedResult, robSpace(testCase2));
        assertEquals(testCase3ExpectedResult, robSpace(testCase3));
        assertEquals(testCase4ExpectedResult, robSpace(testCase4));
        assertEquals(testCase5ExpectedResult, robSpace(testCase5));
        assertEquals(testCase6ExpectedResult, robSpace(testCase6));
    }
}